### ISSUE TITLE
no-return-await actually has semantic differences

### ISSUE THREAD

**Issue by zenorbi** [palantir/tslint#3933]
The rule "no-return-await" says that awaiting before returning in an async function has no semantic change as to simply returning. This does not hold true if you return in a `finally` block.

```js
function longRunningProcess() {
  return new Promise((resolve) => {
    setTimeout(() => {
      console.log("resolving");
      resolve();
    }, 500);
  });
}

async function test() {
  console.log("===");
  try {
    return longRunningProcess(); // finally block -> resolving -> done
    return await longRunningProcess(); // resolving -> finally block -> done
  } finally {
    console.log("finally block");
  }
}
test().then(() => console.log("done"));
```

Seems like the `finally` only applies to the `longRunningProcess()` function call and not the promise value that the coroutine would resolve.

**Comment by zenorbi** [comment]
This behavior is also true in case of a simple `catch`. If the promise that is being returned rejects, the catch block won't run.

```js
function longRunningProcess() {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      console.log("rejecting");
      reject(new Error("we should catch this"));
    }, 500);
  });
}

async function test() {
  console.log("===");
  try {
    return longRunningProcess(); // finally block -> resolving -> done
    //return await longRunningProcess(); // resolving -> finally block -> done
  } catch(err) {
  	console.log("catch block");
  }
}
test().then(() => console.log("done"));
```

**Comment by ajafff** [comment]
@zenorbi is this a bug report or are you just missing these details in the rule's documentation?
AFAICT the rule should leave the cases mentioned above alone, because of these semantic differences.

**Comment by zenorbi** [comment]
@ajafff Just checked and the rule works correctly, so this isn't really a bug report, but a documentation update request. The only thing this begs for now, is an option to report cases where we should use return await. Should I open a new ticket for that?

**Comment by ajafff** [comment]
> The only thing this begs for now, is an option to report cases where we should use return await.

Such a rule already exists, but not as part of TSLint.
I started developing my own linter quite some time ago. It's called [Fimbullinter](https://github.com/fimbullinter/wotan). It has a [core rule](https://github.com/fimbullinter/wotan/tree/master/packages/mimir#rules) `try-catch-return-await` that checks exactly what you are requesting.

---

The only downside is that you need to use two linting tools... Or you could use Fimbullinter to execute your TSLint rules: [plugin 1](https://github.com/fimbullinter/wotan/blob/master/packages/valtyr/README.md) or [plugin 2](https://github.com/fimbullinter/wotan/blob/master/packages/heimdall/README.md).

Another alternative: you can make Fimbullinter rules available to TSLint:

```ts
// ./rules/tryCatchReturnAwaitRule.ts
import { Rule } from '@fimbul/mimir/src/rules/try-catch-return-await';
import { wrapRuleForTslint } from '@fimbul/bifrost';

const Wrapped = wrapRuleForTslint(Rule);
export { Wrapped as Rule };
```

Then compile this file together with your other sources (or use `ts-node`, see TSLint docs).
And finally add this rule to your `tslint.json`

```js
{
  "rulesDirectories": "./dist/rules", // or wherever your compiled files are located
  "rules": {
    // your other rule configurations go here ...
    "try-catch-return-await": true
  }
}
```

**Comment by shusson** [comment]
Yeah just ran into a situation where our code was not handling an error because we omitted the `await`. I think this rule should be reversed so that `await` is always enforced.

**Comment by sebastiendan** [comment]
I also just faced such an issue. A method returning a `Promise.all([...].map(async () => {}))` was not giving the hand back to the caller, while returning `await Promise.all(...)` works fine.

**Comment by zuozp8** [comment]
also with async stack traces there is difference
on node12
```js
const sleep = () => new Promise((r) => setTimeout(r, 10))

async function x() {
    await sleep()
    throw new Error('foo')
}

async function withAwait() {
    await sleep()
    return await x()
}

async function withoutAwait() {
    await sleep()
    return x()
}

;(async() => {
    try {
        await withAwait()
    } catch (e) {
        console.log(e.stack)
    }
    await sleep()
    try {
        await withoutAwait()
    } catch (e) {
        console.log(e.stack)
    }
})()
```
prints
```
Error: foo
    at x (/home/konrad/.PhpStorm2019.2/config/scratches/scratch.js:5:11)
    at async withAwait (/home/konrad/.PhpStorm2019.2/config/scratches/scratch.js:10:12)
    at async /home/konrad/.PhpStorm2019.2/config/scratches/scratch.js:20:9
Error: foo
    at x (/home/konrad/.PhpStorm2019.2/config/scratches/scratch.js:5:11)
    at async /home/konrad/.PhpStorm2019.2/config/scratches/scratch.js:26:9
```

i would even like rule like `force-return-await`

**Comment by JoshuaKGoldberg** [comment]
☠️ *TSLint's time has come!* ☠️ 

TSLint is no longer accepting most feature requests per #4534. See **[typescript-eslint.io](https://typescript-eslint.io)** for the new, shiny way to lint your TypeScript code with **ESLint**. ✨ 

It was a pleasure open sourcing with you all!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
